core/src/protocols/stream/quic/parser.rs

```diff
@@ -0,0 +1,292 @@
+            }
+            let scid_bytes = &data[scid_start..scid_start + scid_len as usize];
+            let scid = Quic::vec_u8_to_hex_string(scid_bytes);
+
+            // Counts all bytes remaining
+            let payload_bytes_count = data.len() - scid_start - scid_len as usize;
+            Ok(Quic {
+                payload_bytes_count: payload_bytes_count as u16,
+                short_header: None,
+                long_header: Some(QuicLongHeader {
+                    packet_type,
+                    type_specific,
+                    version,
+                    dcid_len,
+                    dcid,
+                    scid_len,
+                    scid,
+                }),
+            })
+        } else {
+            // Short Header
+            let mut max_dcid_len = 20;
+            if data.len() < 1 + max_dcid_len {
+                max_dcid_len = data.len() - 1;
+            }
+            let dcid_bytes = data[1..1 + max_dcid_len].to_vec();
+            // Counts all bytes remaining
+            let payload_bytes_count = data.len() - 1 - max_dcid_len;
+            Ok(Quic {
+                short_header: Some(QuicShortHeader {
+                    dcid: None,
+                    dcid_bytes,
+                }),
+                long_header: None,
+                payload_bytes_count: payload_bytes_count as u16,
+            })
+        }
+    }
+}
+
+impl QuicParser {
+    fn check_connection_id(&self, dcid_bytes: &[u8]) -> Option<String> {
+        let dcid_hex = Quic::vec_u8_to_hex_string(dcid_bytes);
+        for dcid_len in (1..dcid_bytes.len() + 1).rev() {
+            let dcid = &dcid_hex[..dcid_len * 2];
+            if self.connection_ids.contains(dcid) {
+                return Some(String::from(dcid));
+            }
+        }
+        None
+    }
+
+    fn process(&mut self, data: &[u8]) -> ParseResult {
+        if let Ok(mut quic) = Quic::parse_from(data) {
+            let session_id = self.cnt;
+
+            match &quic.long_header {
+                Some(long_header) => {
+                    if !long_header.dcid.is_empty() {
+                        self.connection_ids.insert(long_header.dcid.clone());
+                    }
+                    if !long_header.scid.is_empty() {
+                        self.connection_ids.insert(long_header.scid.clone());
+                    }
+                }
+                None => {
+                    if let Some(ref mut short_header) = quic.short_header {
+                        short_header.dcid = self.check_connection_id(&short_header.dcid_bytes);
+                    } else {
+                        log::warn!("Malformed packet");
+                        return ParseResult::Skipped;
+                    }
+                }
+            }
+
+            self.sessions.insert(session_id, quic);
+            self.cnt += 1;
+
+            ParseResult::Continue(session_id)
+        } else {
+            ParseResult::Skipped
```

Comment:
Out of scope for this project at this point, but just to note -- another approach would be to have a default version of the QUIC parser that only parses out some QUIC CID-level state (just from long headers?) vs. getting data for every packet. 

For ex, a TLS handshake intentionally only cares about the first few packets in the connection. Or, the Connection record subscribable type parses from every packet but stores only high-level statistics. 

This is out of scope at this stage & also like I said earlier in the quarter, I personally don't know what the "right" default data to parse is for QUIC. It's a tricky design decision for this protocol IMO. I think every packet is totally fine, especially since the most important use-case is filtering / getting data on QUIC!

Reply:
I tried this implementation and one thing that I noticed is that due to remove session being called when a ParseResult::Done occurs, in my implementation right now this also means that the connection id associated with this quic packet also gets removed. This means that all future quic short headers that arrive that would have that connection id get marked as an unknown connection id. Is there a different method that gets called when a connection ends rather than when it is marked ParseResult::Done that I should implement the removal of connection ids under that would resolve this issue? 

I don't know if this is even a significant issue, that a short header is not assigned a connection id, because I'm not sure in what ways it could be used in network analysis beyond the fact it exists. However, I've kept it thus far as a field to allow the option for an end user.

Also on the point of no support for filtering on fields in a Quic packet, is this referring only to connections that are actively running, or is it including when a connection ends?

Reply:
Could you ParseResult::Done on each header (if that is the definition of "session" you want to go with), and only remove the entry in the `sessions` HashMap once the connection naturally terminates or times out?